Add satochip hardware wallet support

Toporin opened the pull request:
This is a pull request to add support for the [Satochip hardware wallet](https://satochip.io/).

The Satochip hardware wallet is based on a jacavard smartcard and is fully open-source.
The wallet is composed of a [javacard applet](https://github.com/Toporin/SatochipApplet) that is to be loaded on the smartcard. An optional 2FA setting allows to confirm transaction requests on a second android device before signing them (iOS version coming soon).

Implementation details:
*Use pysatochip v0.12.5 to interface with the card
*Suport for Satochip applet up to v0.12 (latest version)

New functionalities in Satochip applet v0.12:
*Card authenticity verification based on device certificate & PKI
*Reset to factory option (e.g. if user forget his PIN code or 2FA device)
*Support for importing encrypted seed from a [SeedKeeper](https://seedkeeper.io/)

More info:
https://github.com/Toporin/ (official repository)
https://pypi.org/project/pysatochip/ (pysatochip library)
https://prezi.com/p/mpq-xhh3mxjl/satochip-gent-meetup/ (Slides from previous meetups in Belgium)
https://t.me/Satochip (Telegram support)

Toporin committed:
Add Satochip hardware wallet support (WIP)

Work in progress...

Toporin committed:
Add Dogecoin support for Satochip (+debug)

Toporin committed:
Add NoCardPresent exception

Toporin committed:
Satochip support: clean code

Toporin committed:
use default path for testnet

Toporin committed:
debug: force use of cryptography 41.0.1 (instead of 41.0.2)

Toporin committed:
run build workflow on 'add-satochip-support' branch

Toporin committed:
debug: add cffi

Toporin committed:
debug: dependencies in setup

Toporin committed:
clean setup.py dependencies

For windows build, the order of dependencies appears to affect pip installation.
There is a (minor) conflict for pyscard dependency, as pysatochip requires v>=1.9.9 (resolving to latest 2.0.7), while openpgp requires exactly 2.0.3.
If v2.0.7 is installed before 2.03, it generates an error and some unrelated dependencies (h11, cffi,...) are not installed for an unknown reason.
By simply moving pysatochip up in 'install_requires' list, pyscard v2.0.3 is installed and conflict is avoided...

bitlogik commented:
Thanks for the PR. We need to test that in details. Hope it was not so painful to achieve. We know how hard it can be when it's an external code based (not your design).

First the integration seems mostly complete and pretty clean. Here's the first feedback about this PR based on the Uniblow coding philosophy and practical matters.
- Why limit to EVM ? is is a restriction in the card ? Because in theory the device is a signer, independent from the wallet/chain used ? Is is a starting point? You may ahave missed to remove DOGE, TRX and XTZ if you only need EVM. If so, you can consider using the LEDGER_EVM_LIST, as this is the EVM chains. If DOGE is compatible, what makes it incompatible with BTC and LTC?
- We need to assess the pysatochip dependency tree. The Uniblow philosophy is to be simple and limit the dependencies. We feel that matters for security. KISS and reduced dependencies helps to maintain the wallet security. Of course it is much easier to call for external ready-made libraries, but it should be limited to simple libraries, not bringing a big tree of the hell dependencies tree. See for example how Cryptnox and Ledger were integrated, without requiring any additional dependency.
- Another challenge about the dependency is the compatibility. Uniblow should be compiled and run from source on Windows, Mac, Ubuntu/Debian, and TailsOS.
- Prefer to use the class name rather to add a device_name attribute.
- In the walletside, there should not be any reference to a specific wallet, but only rely (or test) the wallet attribute. This is way more easy to do that in theory than in practice, as in practice the wallet attributes are becoming quite complex. We started to add these, until realizing each wallet is very specific on its own. Anyway, we want to stick on this modular approach, where devices are attributes, like a family of features. For our matter, the attribute "has_screen" is dedicated to wallet that receive the whole transaction (not hashed) because they can analyze it and display to the user. Maybe we can consider that Satochip has a screen, because it needs a smartphone to authorize the transaction ? It would require a more generic eth.add_vrs_satochip(vrs) method (common to Ledger and Satochip). Else a new attribute should be used, eventually only used by Satochip at first.
- Your device seems to reply with a DER signature. This case is fully compatible with Uniblow, and it has all the business logic to handle that. For the record, the LocalFile (ECKeyPair), Cryptnox and OpenPGP (DER converted in Uniblow side) are all DER signature devices. So the majority of devices are using this signature format, and Uniblow handles everything that is required (decoding, "v" recvovery computation). See how these 3 devices are implemented, and this will considerably simplify and lighten the Satochip device integration. For example in this proposal you have some code for EVM "vrs", and some for simple Bitcoin transactiond. But Uniblow modularity precisely avoid this, and separates the wallets (blockchains) from the signer devices. So the device code has no code about the chains specific, it just provides a DER signature, and the wallet part does the job.

bitlogik commented:
More thought about the "has_screen/DER/vrs" issue. In Uniblow, the has_screen device attribute was initially designed for Ledger, in order to send a full message (not hashed), and the wallet expect a special format of reply dedicated to EVM. This is not ideal regarding the wallet/device separation. Because "has_screen" triggers the signature has a format regarding of the blockchain (because Ledger support in Uniblow is only EVM, and it send a signature in a specific format).
Considering all of these, I thnk it's now time to clean this legacy mess and to make something clean. because clearly Satochip challenges this, as it always replies with a DER signature, has no screen, but requires the full transaction to be sent (as far as we understand).

In the case the Satochip card gets a hash to sign (message already hashed), it's the simple plain case, as the other wallet such as OpenPGP (in term of device it converts into DER), Cryptnox and LocalFile devices. Just be isnpired by these device implementation and it should be easy.

If the Satochip expects the full message (not hashed), it's time to clean this specific has_screen designed for Ledger.
has_screen would be kept, but only tells the full message is sent to the card. This can be renamed to sign_plain_message, but note that has_screen is also used for GUI to specify the user she needs to check on her wallet.
Add an attribute vrs_evm, which tells to ETH_wallet that the signature from the signer device is in the form of VRS (and not being DER).
With these 2 device attributes, we could differentiate the Ledger from a Satochip, and let the Satochip sign qith a DER signature, as the other wallets do.

Toporin commented:
Thanks for the detailed review!
I will take a few days to carefully read and implement your suggestions and I will come back with an improved submission!

Toporin commented:
Here are some feedback to the questions and comments in your code review:

* Why limit to EVM ? is is a restriction in the card ? 
The main limitation of the card is that it does not support ed25519 curves, so Solana is not supported. EOS support was not added because I don't know the specifics of this blockchain well enough.
We already have electrum forks that supports BTC and LTC, and Uniblow does not seem to be fully compatible with them, as Uniblow supports only one address. So the user might see one balance with Electrum, and another (smaller) balance with Uniblow, which might be upsetting. So currently we prefer to disable BTC and LTC support for Satochip.

* We need to assess the pysatochip dependency tree. 
The dependencies in pysatochip (v0.12.7) are quite limited:
- pyscard>=1.9.9 
- ecdsa>=0.15
- cryptography>=3.3.2
- certifi
Pyscard and Cryptography are already needed by Uniblow. ECDSA and Certifi are two well maintained, self-contained, cross-platform and very popular libraries used by many projects. The advantage of adding the pysatochip library (as compared to copying the code in Uniblow) is to have a more modular code, more easy to review and to maintain when updates are provided without having to duplicate the same code in multiple locations.

* In the walletside, there should not be any reference to a specific wallet, but only rely (or test) the wallet attribute: 
I tried to make as few changes as possible on the walletside but some behaviours of Satochip are not quite like ledger or devices without screens. Satochip has some kind of screen since an optional feature allows the user to approve signatures on a smartphone application (like a 2FA). 
-For tx signing, we can use has_screen attribute to distinguish Ledger and Satochip from other devices. However, for Doge for example, Satochip uses the double hash of the tx currently (we currently do blind signing for Doge, but this could evolve in the future): even though satochip has the 'has_screen' attribute enabled, it behaves like devices that don't have the has_screen attribute in this case. Also, since DOGE is not supported by the Ledger in Uniblow, I think we can drop the "has_screen" check for Doge?
-For EIP712 signing, Satochip needs the complete message for display on 2FA, but the ledger apparently only needs hashes. This could be solved by providing the full data_obj to the method sign_EIP712() for both satochip and ledger (i.e. any has_screen devices), and computing the hashes (hash_domain, hash_data) in the ledger class as this is Ledger-specific implementation ?
-Idealy, the GUI message for Satochip should also be slightly different than for Ledger ("!!! Once approved, check on your 2FA-device to confirm the signature  !!! "). This could be done by having a specific 'has_screen_message' attribute by device?
- for the vrs/DER issue, we can indeed add a 'use_vrs_evm' attribute to distinguish ledger/satochip devices for evm signatures. But is it more readable than checking the device type directly?

As more features and devices are added in the future, it might become more difficult to use generic attributes to perform signing operations and it might be simpler (and easier to maintain) to perform signing based on the device type?

Please let me know your thoughts!

bitlogik commented:
About the EVM limitation. This is understandable from the Bitcoin family limitation, this wallet is not "full HD" and uses a static key for each blockchain. Also the fact that the Satochip behaves differently depending on the chain (in terms of signature), is not something that Uniblow foreseen as a behavior for the signing part.
Still let me be clear, regarding this wallet architecture, you don't need to understand the details of the blockchains, all the hardware wallet has to do is to provide a signature for a given message. Uniblow is handling everything (chain info, transaction contruction, signature decoding/attaching). Of course if the hardware wallet wants the whole message to sign to be able to analyze in details what's signed, that's an other story, and that would require dedicated program inside the sign device itself. Thats the main reason why the Ledger device in Uniblow is only compatible with EVM, because the Ledger offers one application per chain. So it is like i's a totally different device depending of the chain.
Similarly about EdDSA chains, it assesses the hardware signer capability and only offers/displays the chains that are compatible.
All that being said, I think it would be a first step to start with only EVM chains for the Satochip.

About the libraries. For certifi, it solves some issues in CPython, but we chose not to use it on Windows. For ecdsa, I think you have everything in Uniblow in cryptolib to handle everything.
So I would recommend to be inspired by Cryptnox and OpenPGP devices to write minimum and independent software. For example to send the right APDUs through pycard, and process with cryptolib (returning the DER is OK, will be processed automatically, see next). Not the easiest way for you, but this is the recommended way to add a new hardware wallet. We can assist you on that task.

More technically : use the has_screen attribute, and just expect a DER signature, which is already fully supported by Uniblow. The only issue here is that the expected signature type when "has screen" is specific to the Ledger (which provides info about "v"/parity), and when no screen it expects a DER form. That is not very well design as the "has screen" also tells about the signature form expected. I propose that you change that in the ETHwalet, so that it always expects a DER signature (the add_dignature method specifically work with a standard DER signature). And add a new attribute .provide_parity True in the Ledger device, and ETHwallet processes with add_vrs instead of add_signature if so.
This way it should work straightaway, you just need to query the DER signature from the Satochip and the EVM wallets will work. The same kind of scheme can apply to any others blockchains, but I think the has_screen is not used for all chains for now.

We can add a new optional string attribute in the device, named on_device_check_type, That will be default to "screen" in the base class. In Satochip you put "2FA-device", or "2FA". So it displays nicely to the user.

Let aside the EIP712 for now, we'll see later, one issue at a time. The message being signed is "EIP712_HEADER + hash_domain + hash_data", which is either sent hashed to a standard device, or as it is into a screen enabled device. That is inline with the others signatures. But your device might be specific about this point. We'll find an elegant way to solve this once the other mess are solved.

I think there could be challenges about "card connection" on the Satochip, especially when you need to 2FA the transaction. The Uniblow philosophy on devices is to connect once at startup. So the PIN is only asked once at the beginning, and the session is stayed opened. An idea could be to save the PIN and connect to the card at each command/method (I havent looked at what you done regarding this point).

bitlogik commented:
Looking at your existing code change. The design I propose will considerably simplify the Satochip device part. It will have only the method sign(self, messsage) and return the DER signature. All we need is to separate _has_screen_ from the signature type. So with _has_screen_ the wallet will send the full message to sign, else it sends the hashed one. And with _provide_parity_, it will expect a signature formed with a parity information, else this is a simple DER signature.

The other change in the prepare method, so it only returns tx_bin, and build_tx when no has_screen hashes the message, like it is currently done for BTC wallet. That would be more "logic" and simple.

The code in the ETH wallet  will be like that
```
        tx_bin = self.eth.prepare(account, amount, gazprice, ethgazlimit, data)
        if self.current_device.has_screen:
            if self.eth.contract and self.current_device.ledger_tokens_compat:
                # Token known by Ledger ?
                ledger_info = self.ledger_tokens.get(self.eth.contract.lower())
                if ledger_info:
                    # Known token : provide the trusted info to the device
                    name = ledger_info["ticker"]
                    data_sig = ledger_info["signature"]
                    self.current_device.register_token(
                        name, self.eth.contract[2:], self.eth.decimals, self.chainID, data_sig
                    )
            hash_sign = tx_bin
        else:
            hash_sign = sha3(tx_bin)
        signature = self.current_device.sign(hash_sign)
        if self.current_device.provide_parity:
            tx_signed = self.eth.add_vrs.sign(signature)
        else:
            tx_signed = self.eth.add_signature(signature)
        return tx_signed
```

The others "sign" method in ETH wallet will be likely modified.

Toporin committed:
Includes feedback from Bitlogik

For all devices:
New attribute 'provide_parity': True if device provides a rsv signature with parity information, otherwise a simple DER signature
New attribute 'on_device_check_type': if device 'has_screen', this string describes the type of screen (screen or 2FA device)

When signing EVM tx, provides full data to sign if device has a screen (Ledger or Satochip), and only provides hash otherwise.
Also takes into account the format of the returned signature according to 'provide_parity' attribute (DER signature or rsv)

For Satochip device: always returns DER signature and let Uniblow do the conversion

Toporin committed:
Clean & remove old code

Toporin committed:
Remove 'add_vrs_satochip()' function (not needed anymore)

Toporin commented:
I just modified the PR with your suggestions. 
Let me know if it is going in the right direction.
I still have to remove the pysatochip dependency and include the code to communicate with the satochip inside Uniblow.

bitlogik commented:
Yes, that's definitively goes in the right direction.

Let's start with EVM chains only. So don't touch DOGE.
Still, the changes with wallet properties (such as provide_parity and has_screen), may need to modify the other wallets such as BTC.

We can even simplify the device attribute to a single one for "screen" :
screen = "" < boolean and string, string will be false when empty
The test device.screen attribute, if false, if true, it provides the device check type. And when true, we can rad the string value to display.
Or it can be named to a more generic term  .on_device_check. With this we reduce the number of device attributes.

Possibly, this may be the time to remove from ETHwallet, the messages specific to the device, and put them in app transfer(). That was already stated as _if app.wallet.current_device.has_screen:_ was added in the Uniblow "center" app.
This need to be assesed, ut I just see that the Ledger integration for EVM breaks the modular separation,a dn there are device specific stuff in the wallet part. But we can do that work later. For example the EVM specific user signature (from WalletConnect) are different from the central generic transfer method, and we need that for EVM. So this better separation can be done at la later stage.

Looking at sign_message in devices/Satochip, I can't see any difference with _sign_evm_. The calling wallet part, already add the header needed in the signed message. This needs to be investigated / sorted out.
If we sitck to EVM for the first step of the integration, that means the devices has only one sign method (the device standard one), and get full message to be signed (thanks to the has_screen / on_device_check attribute).

One thinking about the system is the 2FA needs the full message, but the card only needs the hash, with the authorization from 2FA. We may found a new "refactoring" or processing that takes this in account and simplify the integration. I mean if the wallet software has a redesign of some mechanisms, it may be a better for the Satochip wallet. Just a though about this particularity, not sure if helpful in practice.

Please add section in ReadMe to describe the Satochip device, as the others devices listed.
You can add a line Copyright Satochip or Baudoin Collard added in specific files, like devices/Satochip. If you wrote these modules, there nos reason it will be only attributed to BitLogiK.

I have prepared a lot of questions about your wallet that we'll discuss in a private meeting we planned. See you.

Toporin committed:
WIP: remove pysatochip library and implements functionalities natively in Uniblow

done: commented unused code

Toporin committed:
remove ecc.py dependency

Use cryptolib.cryptography: public_key_recover() to recover pubkey (in bytes) instead of ECPubkey.from_sig_string(compsig, id, hash)

Toporin committed:
Satochip: remove reference to seedkeeper in version.py

Toporin committed:
satochip: remove reference to ecc.py in CardConnector.py

Toporin committed:
satochip: remove reference to crypto.py in SecureChannel.py

Use cryptolib/cryptography.py: aes_encrypt, aes_decrypt, append_PKCS7_padding, strip_PKCS7_padding instead

Toporin committed:
(minor) improve logs in CardDataParser

Toporin committed:
Remove unused code

Toporin committed:
Use cryptolib.ECKeyPair instead of ecdsa library for ECDH key exchange

Toporin committed:
Use cryptolib.ECKeyPair.ECpubkey to verify signatures, instead of ecdsa

Toporin committed:
Satochip: verify card authenticity

Toporin committed:
Satochip: manage Satochip root certificate

Add Satochip Certificates to uniblow.spec
Remove SeedKeeper, Satodime and test certificates

Toporin committed:
Add Satochip to readme

Toporin committed:
remove certifi dependency

bitlogik committed:
Throw earlier when no Satochip

bitlogik committed:
Clean use of sleep

bitlogik committed:
Code Formatting

bitlogik committed:
Satochip button update

bitlogik committed:
More flexible on OpenPGP responses

bitlogik commented:
Lot of work done, so the integration is better, as it is done in the spirit of this software. About integration, I think lot of functions in utils can found their counterpart in cryptolib. I think about all the hash functions, and the base58 conversions.

I performed some tests, and the biggest issue is with the card observer. The main issue is when a user clicks on the Satochip device button right after inserting the card, she needs to wait for the insertion observer to finish. That is quite a long time. And the effect is that the app does nothing (but wait) during a second when one selects the Satochip device.
Also the observers are triggered/started after one clicks on Satochip device button. Then they are triggered whenever the user insert of remove a smartcard/dongle. So these observers are really annoying, bring some issues, and I'm not sure why they are needed or used for here. I acknowledge it could be useful to have a removal oberver to detect when the card is not present. But at least it should be only triggered when a Satochip is in use, as the active device.

On the code I performed the following :

- Change the button image with the right template
- Code formatting
- Rebase on master -> change branch to "add-satochip-rebase"

The last point generated a lot of conflicts, and I'm not sure the resolution was OK. That requires a lot of tests of manual transactions to confirm everything is still OK. The automated tests pass, but they are mostly focused on internal computations.
Since there was so much conflicts and I'm not sure about the result, I created a new branch named "add-satochip-rebase". This needs lot of testing to confirm.

Code cleaning must continue. The pysatochip dependency can be removed? I'm not sure about the Apache v2 license in some files, I would prefer only one license across all files.

Toporin commented:
Ok, I will look what I can do regarding the following:
* util.py: see what can be removed and use cryptolib functions instead
* Try to remove cardObserver
* change apache license to GPL

The pysatochip dependency can indeed be removed.

I will work on these points this week and beginning of next week and let you know when changes are ready.

Toporin committed:
remove pysatochip dependency

Toporin committed:
Update license for Satochip

Toporin committed:
Rename certificate_validator.py to CertificateValidator.py to follow project convention

Toporin committed:
Update license for JCconstants.py in Satochip

Toporin committed:
Use cryptolib methods instead util.py methods

todo: clean code

Toporin committed:
Remove util.py (deprecated)

Toporin committed:
Satochip: remove CardObserver class in CardConnector

Toporin committed:
Satochip: refactor card_verify_PIN and PIN management

Toporin commented:
I have updated the PR starting from your latest changes:
* fully remove pysatochip dependency
* removed util.py and use cryptolib functions instead
* removed cardObserver
* use same license as Uniblow
* Refactor and clean more code

bitlogik committed:
Add FDUSD token

bitlogik committed:
Code format

bitlogik committed:
Fix bugs from variables typo

bitlogik committed:
Code quality

bitlogik commented:
Rebased on latest master. Then some cleaning commits were added.
This seems ready to be merged. We just need to perform many tests to confirm that everything still works. Can you send me an email to plan a meeting for this matter?

Toporin commented:
Here is my email: satochip.wallet@gmail.com
I am available this wednesday morning, Thursday or Friday (mostly the whole day).
Next week will be a bit more challenging as it's a school holiday week in Belgium :-)

Toporin committed:
Fix Dogecoin tx error: 'LocalFile' object has no attribute 'device_name'

Full error:
Error during the transaction processing : 'LocalFile' object has no attribute 'device_name'
Traceback (most recent call last):
  File "uniblow.py", line 685, in perform_transfer
    tx_info = app.wallet.transfer(amnt, to, fees)
  File "/home/satochip/Documents/github/uniblow/wallets/DOGEwallet.py", line 303, in transfer
    return self.raw_tx(shift_10(amount, DOGE_units), fee, to_account)
  File "/home/satochip/Documents/github/uniblow/wallets/DOGEwallet.py", line 287, in raw_tx
    if self.current_device.device_name == "Satochip":
AttributeError: 'LocalFile' object has no attribute 'device_name'
